squid-forward-proxy-operator/src/http_proxy.py

```diff
@@ -0,0 +1,882 @@
+
+    def get(self, requirer_id: str) -> Optional[HttpProxyRequest]:
+        """Get a specific HTTP proxy request.
+
+        Args:
+            requirer_id: requirer id of the proxy request.
+
+        Returns:
+            the proxy request.
+        """
+        if requirer_id not in self._requests:
+            return None
+        request = copy.deepcopy(self._requests[requirer_id])
+        request["group"] = self._integration_id
+        if not request.get("src_ips"):
+            src_ips = self._get_remote_unit_ips()
+            request["src_ips"] = src_ips
+            request["implicit_src_ips"] = True
+        else:
+            request["implicit_src_ips"] = False
+        request["id"] = request["requirer"]
+        del request["requirer"]
+        return HttpProxyRequest(**request)
+
+
+class HttpProxyRequestListReadWriter(HttpProxyRequestListReader):
+    """Integration helper: read and write request list."""
+
+    @property
+    def _integration_data(self) -> ops.RelationDataContent:
+        """Get integration data.
+
+        Returns:
+            integration data.
+        """
+        return self._integration.data[self._charm.app]
+
+    def _dump(self) -> None:
+        """Write HTTP requests in the buffer to the integration."""
+        requests = [self._requests[id] for id in sorted(self._requests)]
+        self._integration_data["requests"] = json.dumps(
+            requests, sort_keys=True, ensure_ascii=True
+        )
+
+    def delete(self, requirer_id: str) -> None:
+        """Delete a HTTP proxy request.
+
+        Args:
+            requirer_id: requirer id of the proxy request.
+        """
+        self._requests.pop(requirer_id, None)
+        self._dump()
+
+    def add(
+        self,
+        requirer_id: str,
+        domains: list[str],
+        auth: list[str],
+        src_ips: list[str] | None = None,
+    ) -> None:
+        """Add a new HTTP proxy request.
+
+        Args:
+            requirer_id: requirer id of the proxy request.
+            domains: proxy request domains.
+            auth: proxy request auth.
+            src_ips: proxy request src_ips.
+
+        Raises:
+            KeyError: request already exists.
+        """
+        if requirer_id in self._requests:
+            raise KeyError(
+                f"http proxy request with requirer id {repr(requirer_id)} already exists"
+            )
+        # here to validate the inputs only
+        HttpProxySpec(
+            group=0,
+            id=requirer_id,  # type: ignore
+            domains=domains,  # type: ignore
+            auth=auth,  # type: ignore
+            src_ips=src_ips or ["10.0.0.1"],  # type: ignore
+        )
+        request = {
+            "requirer": requirer_id,
+            "domains": domains,
+            "auth": auth,
+        }
+        if src_ips:
+            request["src_ips"] = src_ips
+        self._requests[requirer_id] = request
+        self._dump()
+
+
+class HttpProxyResponseListReader:
```

Comment:
Should be a private class

Reply:
Updated along with other three classes, thanks!